modules/data.remote/R/GEDI_AGB_prep.R

```diff
@@ -56,481 +85,467 @@ GEDI_AGB_prep <- function(site_info, time_points, outdir = file.path(getwd(), "G
+  i <- NULL
+  agb_mean_sd <- foreach::foreach(i = seq_along(which.point.in.which.file), .packages = c("Kendall", "purrr"), .options.snow=opts) %dopar% {
+    temp.files <- which.point.in.which.file[[i]]
+    # if we only have 1 or 0 files that cover this pixel.
+    if (length(temp.files) <= 1) {
+      return(list(agb_mean = NA, agb_sd = NA))
+    }
+    file.keep.dat <- vector("list", length = length(temp.files))
+    # loop over files.
+    for (j in seq_along(temp.files)) {
+      # load file.
+      level4a_h5 <- hdf5r::H5File$new(temp.files[j], mode = "r")
+      # load beams id.
+      groups_id <- grep("BEAM\\d{4}$", gsub("/", "", hdf5r::list.groups(level4a_h5,recursive = F)), value = T)
+      dat <- c()
+      for (beam in groups_id) {
+        level4a_i <- level4a_h5[[beam]]
+        # if we have that datasets for the specific beam.
+        if (any(hdf5r::list.datasets(level4a_i) == "shot_number")) {
+          # grab lat/lon, agb, and the orbit information.
+          rhs <- data.table::data.table(lat_lowestmode = level4a_i[["lat_lowestmode"]][],
+                                        lon_lowestmode = level4a_i[["lon_lowestmode"]][],
+                                        agbd = level4a_i[["agbd"]][],
+                                        orbit = substr(level4a_i[["shot_number"]][], 1, 5))
+          xvar <- t(level4a_i[["xvar"]][,]) %>% `colnames<-`(c("RH25", "RH50", "RH75", "RH98"))
+          predict_stratum <- level4a_i[["predict_stratum"]][]
+          rhs <- cbind(rhs, xvar, predict_stratum)
+          # grab records that have usable agb estimations.
+          rhs <- rhs[which(rhs$agbd>=0),]
+          dat <- rbind(dat, rhs)
+        }
       }
+      # filtering data by the bounding box.
+      diff.lat <- abs(site_info$lat[i] - dat$lat_lowestmode)
+      diff.lon <- abs(site_info$lon[i] - dat$lon_lowestmode)
+      keep.inds <- which(diff.lat <= buffer & diff.lon <= buffer)
+      file.keep.dat[[j]] <- list(dat = dat[keep.inds, ])
     }
-    #time period
-    daterange <- c(from, to)
-    
-    # Get path to GEDI2A data
-    gLevel4 <-
-      GEDI4R::gedifinder(
-        ul_lat,
-        ul_lon,
-        lr_lat,
-        lr_lon,
-        daterange = daterange
-      )
-    
-    lg <- length(gLevel4)
-    
-    if(lg==0){stop("there are no GEDI files for this date or coordinates")}
-    
-    if (just_path) {
-      return(gLevel4)
-      stop(invisible())
+    # grab ancillary table.
+    ancillary <- level4a_h5[["ANCILLARY"]][["model_data"]]
+    vcov <- array(unlist(lapply(1:35,function(x)ancillary[x][["vcov"]])), dim = c(5, 5, 35))
+    predict_stratum <- do.call(rbind,lapply(1:35,function(x)ancillary[x][["predict_stratum"]]))
+    # combining data and extract records that have the most abundant predict stratum.
+    dat <- file.keep.dat %>% purrr::map(function(d){d$dat}) %>% dplyr::bind_rows()
+    most_stratum <- names(sort(table(dat$predict_stratum), decreasing = TRUE))[1]
```

Comment:
Nice refactor to eliminate rhdf5, and I see why it would have been nicer to use it if the installation had cooperated! 😀